lucene/facet/src/java/org/apache/lucene/facet/taxonomy/directory/DirectoryTaxonomyReader.java

```diff
@@ -323,8 +323,10 @@ public FacetLabel getPath(int ordinal) throws IOException {
       }
     }
     
-    Document doc = indexReader.document(ordinal);
-    FacetLabel ret = new FacetLabel(FacetsConfig.stringToPath(doc.get(Consts.FULL)));
+    boolean found = MultiDocValues.getBinaryValues(indexReader, Consts.FULL).advanceExact(catIDInteger);
```

Comment:
I think instead of the boxed `Integer catIDInteger` we should pass the `int ordinal` to `.advanceExact(...)`?  Not the cause of the `NPE`, just cleaner.

Reply:
OK, I see one issue -- you are pulling a new `BinaryDocValues`, calling `.advanceExact` on it (good), but then pulling a new `BinaryDocValues` below and not calling `.advanceExact` on it.

I think you must add a new local variable, e.g. `BinaryDocValues values`.  Pull it once (using the `MultiDocValues.getBinaryValues` sugar API).  Then call `.advanceExact` on that and assert it succeeded. Finally, use that same `values` instance (now that it has advanced to the right `docId`) to call `.binaryValue().utf8ToString()`.

I think that should fix the `NPE`?

This is misuse of the API for the default Lucene Codec for `BinaryDocValues`, since you were calling `.binaryValue()` before `.advanceExact()`.  It is somewhat disappointing that the codec threw a confusing `NPE` and not a clearer (best effort) exception stating that you must first call `.advanceExact`.  Maybe we could improve the default Codec?  (Though, not if that would hurt performance of correct usage).  OK I see: the `NPE` is because of `MultiDocValues.currentValues` is `null` since `.advanceExact` was not yet called.  Maybe we could add an `assert` there, confirming `.advanceExact` was indeed called and had returned `true`?  It would have made debugging this easier, and should not hurt performance when assertions are disabled ...

Reply:
One more idea: instead of using `MultiDocValues` sugar, I think we should use Lucene's `ReaderUtil` to quickly (binary search) determine which leaf holds this `docId`, then pull `BinaryDocValues` from that `LeafReader`?

Reply:
Thank you for looking at it so closely (and helping in debugging).

The new PR has the following changes:
1. Use `ordinal`  instead of `catIDInteger` (IntelliJ says that boxing is anyways not needed, perhaps we can remove?)
2. Use the correct `values` instance that has advanced to the correct `docId`
3. Use `ReaderUtil` to get to the `leaf` and then use `LeafReader` instead of using the higher level `MultiDocValues` call

TEST:
`ant test` in the `lucene-solr/lucene/facets` directory (passes successfully)
`ant precommit`

I've not added any new tests because this PR changes a low level implementation detail and the current tests already cover this

The next step is to run lucene benchmarks!